### PR TITLE
Add namespace support on PodTemplate.

### PR DESCRIPTION
**Pull request by iocanel**
This pull request, adds support for namespaces inside the PodTemplates.

This feature can be really useful, if we need to create `ephemeral` namespace to run our builds and tests. For example: 

- create a namespace 
- deploy all apps and services 
- run end to end tests 
- then destroy namespace.

### CODE REVIEW

**Comment by iocanel** [comment]
Some more background: I am experimenting on porting [arquillian-cube-kubernetes](https://github.com/arquillian/arquillian-cube/blob/master/docs/kubernetes.adoc) into a jenkins plugin, so that it can be used to include non-java end-to-end tests from within jenkins pipelines.

The missing piece is being able to sping `agent pods` in other namespaces.

File: README.md [hunk 1]
```diff
@@ -162,7 +163,13 @@ Then consumers of the library could just express the need for a maven pod with d
             """
         }
     }
+
+#### Using a different namespace
  
+There might be cases, where you need to have the slave pod run inside a different namespace than the one jenkins is running in.
```
Comment (carlossg) [inline README.md]: IIRC pods are started in the namespace in the cloud definition, not in the one jenkins is running

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 2]
```diff
@@ -66,18 +67,19 @@
...
         this.containerName = containerName;
         this.alive = alive;
         this.started = started;
         this.finished = finished;
     }
 
     @Deprecated
-    public ContainerExecDecorator(KubernetesClient client, String podName, String containerName, String path, AtomicBoolean alive, CountDownLatch started, CountDownLatch finished) {
-        this(client, podName, containerName, alive, started, finished);
+    public ContainerExecDecorator(KubernetesClient client, String podName, String namespace, String containerName, String path, AtomicBoolean alive, CountDownLatch started, CountDownLatch finished) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: this is deprecated, shouldn't be touched

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java [hunk 3]
```diff
@@ -66,18 +67,19 @@
     private transient ExecWatch watch;
     private transient ContainerExecProc proc;
 
-    public ContainerExecDecorator(KubernetesClient client, String podName, String containerName, AtomicBoolean alive, CountDownLatch started, CountDownLatch finished) {
+    public ContainerExecDecorator(KubernetesClient client, String podName, String namespace, String containerName, AtomicBoolean alive, CountDownLatch started, CountDownLatch finished) {
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: changing public signature, either have a constructor without args or deprecate the old one and add a new one

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java [hunk 4]
```diff
@@ -137,6 +137,7 @@ public static PodTemplate combine(PodTemplate parent, PodTemplate template) {
 
         PodTemplate podTemplate = new PodTemplate();
         podTemplate.setName(name);
+        podTemplate.setNamespace(!Strings.isNullOrEmpty(template.getNamespace()) ? template.getNamespace() : parent.getNamespace());
```
Comment (carlossg) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java]: will this get the namespace from the cloud definition if not set?

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java (hunk 3 above)
Comment (iocanel) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/pipeline/ContainerExecDecorator.java]: Will do!

File: src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java (hunk 4 above)
Comment (iocanel) [inline src/main/java/org/csanchez/jenkins/plugins/kubernetes/PodTemplateUtils.java]: passing a null namespace to the client, result in using whatever is already configured and this this case the one from the cloud definition.

**Comment by iocanel** [comment]
@carlossg: if we just add a new constructor and deprecate the old we are hitting issues, due to having multiple constructors with the same signature. This is a result of removing a String proerty called `path`, and later on adding an other String property called `namespace`.

**Comment by iocanel** [comment]
@carlossg: Applied all your suggestions and also added a test to verify that overriding the namespace in the pipeline works without issues.

**Pull request merged**